### PR TITLE
Add example sites and calculate API weights

### PR DESCRIPTION
**Pull request by jdorweiler**
**Adds example sites for each request**
I'm adding a list of example sites where a request was found on while crawling. You can define a the number of sites to include and a list of top sites look for in the config file. If you include a top site list the example sites will be made up of half top sites (if any were found). 

**Recalculate API weights for fingerprint scoring**
While processing crawler data we'll keep track of the number of scripts that are considered tracking (uses 15 or more APIs) and non-tracking, and the APIs they're using. 
API weights are calculated as:
```
(# tracking scripts API was used on) / (# tracking scripts) / (# non-tracking scripts API was used on) / (# non-tracking scripts)
```
If an API was only used by tracking scripts then it defaults to the highest weight we have seen from any other API.

### CODE REVIEW

File: src/trackers/classes/crawl.js [hunk 1]
```diff
@@ -95,6 +101,29 @@ function _processSite (crawl, site) {
             crawl.commonRequests[key].update(request, site)
         }
     })
+
+    for (const [script, apis] of Object.entries(site.siteData.data.apis.callStats)) {
+        const apisUsed = Object.keys(apis)
+        const tracking = apisUsed.length >= 15 ? true : false
```
Comment (kdzwinel) [inline src/trackers/classes/crawl.js]:
```suggestion
        const tracking = (apisUsed.length >= API_COUNT_THRESHOLD)
```

Can we export threshold to a separate variable somewhere at the top?

File: src/trackers/classes/crawl.js [hunk 2]
```diff
@@ -131,6 +160,35 @@ function _getDomainSummaries (crawl) {
     return domainSummary
 }
 
+function _getFingerprintWeights (crawl) {
```
Comment (kdzwinel) [inline src/trackers/classes/crawl.js]: can we extract that as a helper? it'd be easier to test.

File: src/trackers/classes/rule.js [hunk 3]
```diff
@@ -12,6 +13,48 @@ class Rule {
...
+            this.exampleSites = _getExampleSites(newRuleData.pages, sharedData.config.includeExampleSites)
+        }
+    }
+}
+
+/* Adds example sites for each rule if enabled in the config file
+ * Half of the sites can be chosen from a top sites list provided in the config. 
+ * The other half are randomly chosen form the list of sites a request was found on
+ */
+function _getExampleSites (sites, limit) {
```
Comment (kdzwinel) [inline src/trackers/classes/rule.js]: ditto, lets extract to a helper

File: src/trackers/helpers/sharedData.js [hunk 4]
```diff
@@ -33,6 +34,14 @@ function _readEntities () {
     }, {})
 }
 
+// option list of top example sites to include in tracker files
+function _getTopExampleSites (config) {
+    if (!config.topExampleSites) {
+        return null
+    }
+    return new Set(JSON.parse(fs.readFileSync(config.topExampleSites, 'utf8')))
```
Comment (kdzwinel) [inline src/trackers/helpers/sharedData.js:42]: can we document how this file should look like (structure)?

**Review by kdzwinel — COMMENTED** [review]
I did a quick first pass and left some comments. There are also some linting issues.

File: src/trackers/classes/crawl.js (hunk 2 above)
Comment (jdorweiler) [reply src/trackers/classes/crawl.js]: 👍

File: src/trackers/helpers/sharedData.js (hunk 4 above)
Comment (jdorweiler) [reply src/trackers/helpers/sharedData.js:42]: Added a comment to the config file

File: src/trackers/helpers/fingerprints.js [hunk 5]
```diff
@@ -0,0 +1,55 @@
...
+    if (!fp) {return 0}
+
+    const logFp = Math.log(fp)
+    let bucket = 0
+
+    if (logFp > 6) {
+        bucket = 3
+    } else if (logFp > 3.5 && logFp <= 6) {
+        bucket = 2
+    } else if (logFp <= 3.5) {
```
Comment (kdzwinel) [inline src/trackers/helpers/fingerprints.js:14]: those magic values, similarly to api threshold, is something that we should probably also update automatically in the long run, right?

File: src/trackers/helpers/rule.js [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    const exampleSites = idxList.map(idx => sites[idx])
+
+    return exampleSites
+}
+
+// get unique random indicies from the source list
+function _getRandomIndex (sourceList, indexList) {
+    const idx = Math.floor(Math.random() * sourceList.length)
+
+    // skip sites already in the example list, or IPs
```
Comment (kdzwinel) [inline src/trackers/helpers/rule.js]: "or IPs" part is outdated, right?

File: src/trackers/helpers/rule.js [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    if (sites.length <= limit) {
+        return sites
+    }
+
+    const idxList = []
+
+    // half of the list is topExampleSites if given
+    if (sharedData.topExampleSitesSet) {
+        const intersection = sites.filter(x => sharedData.topExampleSitesSet.has(x))
+        intersection.slice(0, Math.floor(limit/2)).map(domain => idxList.push(sites.indexOf(domain)))
```
Comment (kdzwinel) [inline src/trackers/helpers/rule.js]:
```suggestion
        sites
            .filter(x => sharedData.topExampleSitesSet.has(x))
            .slice(0, Math.floor(limit/2))
            .forEach(domain => idxList.push(sites.indexOf(domain)))
```

nit: removed variable as we don't need it, map->forEach, formatting

File: src/trackers/helpers/fingerprints.js [hunk 8]
```diff
@@ -0,0 +1,55 @@
...
+
+    for(const [api, apiScores] of Object.entries(crawl.fpWeights.apis)) {
+        const trackingWt = apiScores.tracking / crawl.fpWeights.scripts.tracking
+        const nontrackingWt = apiScores.nontracking / crawl.fpWeights.scripts.nontracking
+
+        // api is always used for tracking, set to highest weight we see
+        if (nontrackingWt === 0) {
+            setMaxWeight.push(api)
+        } else {
+            const weight = (trackingWt / nontrackingWt).toFixed(2)
```
Comment (kdzwinel) [inline src/trackers/helpers/fingerprints.js:34]: https://github.com/duckduckgo/tracker-radar/blob/jd/fp-weights/build-data/generated/api_fingerprint_weights.json still has weights that are not `.toFixed(2)`, but I'm guessing that that file is just old?

**Review by kdzwinel — COMMENTED** [review]
I left couple of small comments, but feel free to merge once you address them.

File: src/trackers/helpers/fingerprints.js (hunk 5 above)
Comment (jdorweiler) [reply src/trackers/helpers/fingerprints.js:14]: Yes eventually

File: src/trackers/helpers/fingerprints.js (hunk 8 above)
Comment (jdorweiler) [reply src/trackers/helpers/fingerprints.js:34]: Yes, I updated this after building that data.

**Pull request merged**